### ISSUE TITLE
Add support for Hibernate 6.4

### ISSUE THREAD

**Issue by vladmihalcea** [vladmihalcea/hypersistence-utils#685]
When testing the `hypersistence-utils-hibernate-63` module with Hibernate 6.4, the following issue is thrown:

`Caused by: java.lang.ClassCastException: class org.hibernate.type.CustomType cannot be cast to class org.hibernate.type.BasicPluralType (org.hibernate.type.CustomType and org.hibernate.type.BasicPluralType are in unnamed module of loader 'app')`

Either the Types in the `hypersistence-utils-hibernate-63` module need to be changed or a new module will have to be created for this issue to be fixed.

**Comment by cmuchinsky** [comment]
One comment regarding the deprecation of `PostgreSQLEnumType` [here](https://github.com/vladmihalcea/hypersistence-utils/blob/master/hypersistence-utils-hibernate-62/src/main/java/io/hypersistence/utils/hibernate/type/basic/PostgreSQLEnumType.java#L31) and its removal from `hypersistence-utils-hibernate-63` in favor of Hibernate's new `PostgreSQLEnumJdbcType` class. It appears there is a [bug](https://hibernate.atlassian.net/browse/HHH-17496) with the new Hibernate implementation that might prevent it from behaving the same since it defaults to the Enum's `toString()` instead of `name()` in some situations.

**Comment by vladmihalcea** [comment]
@cmuchinsky It's surely possible to create a custom Type for Hibernate 6.3 and 6.4 that retains the old behavior while preventing that bug. You just have to re-adapt the old Type to the new API changes that were done in latest Hibernate versions.

**Comment by cmuchinsky** [comment]
> @cmuchinsky It's surely possible to create a custom Type for Hibernate 6.3 and 6.4 that retains the old behavior while preventing that bug. You just have to re-adapt the old Type to the new API changes that were done in latest Hibernate versions.

Indeed, I have opened a [PR](https://github.com/hibernate/hibernate-orm/pull/7643) so it hopefully gets fixed directly in Hibernate, I just wanted to raise awareness mostly for the community in case they read this.

**Comment by vladmihalcea** [comment]
You can create that Type either in Hibernate or externally, like in this library. If that PR is not integrated, you can just adapt the old Hypersistence Utils PostgreSQLEnumType for Hibernate 6.4.

**Comment by vladimirfx** [comment]
This hits me - I can't update to Spring Boot 3.2.1 (which contains important fixes). Can I help in any way to resolve the class cast problem?

**Comment by vladmihalcea** [comment]
You need to fork the project and investigate how to fix it. Send me a Pull Request when you are done.

**Comment by adrianAzoitei** [comment]
Looking into it too on https://github.com/vladmihalcea/hypersistence-utils/pull/691

**Comment by vladimirfx** [comment]
While investigating I've found Hibernate native array support for basic fields.  See `BasicCollectionType` and `ConvertedBasicCollectionType`. 

I think `hypersistence-utils-hibernate-64` can just drop array support in favor of Hibernate native approach. Implementing a backward-compatible variant of `AbstractArrayType` is somehow expensive because it requires determining the array element type at configuration time.

**Comment by adrianAzoitei** [comment]
Started a topic on Hibernate's Forum: https://discourse.hibernate.org/t/collection-usertypes-cause-a-cast-exception-in-6-4/8752

**Comment by vladmihalcea** [comment]
Great. Hopefully Hibernate fixes the issue.

**Comment by dmitry-weirdo** [comment]
Is this related to this exception?

```java
Caused by: java.lang.NoClassDefFoundError: org/hibernate/type/descriptor/converter/spi/EnumValueConverter
	at io.hypersistence.utils.hibernate.type.HibernateTypesContributor.contribute(HibernateTypesContributor.java:61) ~[hypersistence-utils-hibernate-62-3.3.2.jar:?]
	at org.hibernate.boot.internal.MetadataBuilderImpl.applyTypes(MetadataBuilderImpl.java:295) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[?:?]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyTypeContributors(EntityManagerFactoryBuilderImpl.java:1404) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyMetamodelBuilderSettings(EntityManagerFactoryBuilderImpl.java:1385) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:280) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:198) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:63) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:376) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:352) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1820) ~[spring-beans-6.1.2.jar:6.1.2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1769) ~[spring-beans-6.1.2.jar:6.1.2]
	... 96 more
Caused by: java.lang.ClassNotFoundException: org.hibernate.type.descriptor.converter.spi.EnumValueConverter
	at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at io.hypersistence.utils.hibernate.type.HibernateTypesContributor.contribute(HibernateTypesContributor.java:61) ~[hypersistence-utils-hibernate-62-3.3.2.jar:?]
	at org.hibernate.boot.internal.MetadataBuilderImpl.applyTypes(MetadataBuilderImpl.java:295) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092) ~[?:?]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyTypeContributors(EntityManagerFactoryBuilderImpl.java:1404) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyMetamodelBuilderSettings(EntityManagerFactoryBuilderImpl.java:1385) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:280) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:198) ~[hibernate-core-6.4.1.Final.jar:6.4.1.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:63) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:376) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:352) ~[spring-orm-6.1.2.jar:6.1.2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1820) ~[spring-beans-6.1.2.jar:6.1.2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1769) ~[spring-beans-6.1.2.jar:6.1.2]
	... 96 more
```

**Comment by vladmihalcea** [comment]
No. Your exception is because you are trying to use the 62 module with Hibernate 6.4 or 6.3.

**Comment by dmitry-weirdo** [comment]
> No. Your exception is because you are trying to use the 62 module with Hibernate 6.4 or 6.3.

Ook, should I then use https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-60 instead of https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-52?

**Comment by vladmihalcea** [comment]
> Ook, should I then use https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-60 instead of https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-52?

No. You need to upgrade to `hypersistence-utils`, as explained in the [README](https://github.com/vladmihalcea/hypersistence-utils?tab=readme-ov-file#installation-guide). There's no `hibernate-types` anymore.

**Comment by dmitry-weirdo** [comment]
> > Ook, should I then use https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-60 instead of https://mvnrepository.com/artifact/com.vladmihalcea/hibernate-types-52?
> 
> There's no `hibernate-types` anymore.
> 
> You need to upgrade to `hypersistence-utils`, as explained in the [README](https://github.com/vladmihalcea/hypersistence-utils?tab=readme-ov-file#installation-guide) if you plan on using Hibernate 6.3 or a lower version.
> 
> But, if you want to use `hibernate-core-6.4.1`, as indicated by your error stack trace, then you only option is to fix this issue.

You're correct. Sorry, I copied the artifact from the old branch.

I switched from `6.2` to

```xml
            <dependency>
                <groupId>io.hypersistence</groupId>
                <artifactId>hypersistence-utils-hibernate-63</artifactId>
                <version>3.7.0</version>
            </dependency>
```
But there is no `io.hypersistence.utils.hibernate.type.basic.PostgreSQLEnumType;` anymore?

**Comment by vladmihalcea** [comment]
> But there is no io.hypersistence.utils.hibernate.type.basic.PostgreSQLEnumType anymore?

No, there is not. Instead, you can use Hibernate native alternative:

    @JdbcType(PostgreSQLEnumJdbcType::class)

**Comment by jmax01** [comment]
PG Enums with native queries are an issue.  I will create a separate ticket for this and our solution.

**Comment by reneleonhardt** [comment]
As soon as I add `@Type(ListArrayType.class)` to an entity I get the same CustomType / BasicPluralType exception with
Spring Boot 3.2.1 / Hibernate 6.4.1.
When I downgrade to `<hibernate.version>6.3.2.Final</hibernate.version>` I can read and write this entity without problems.
All 14 ArrayType classes show the same exception, all Hibernate 6 releases contained breaking changes, but 6.4 is especially bad :(

I guess a very simple test would catch this specific exception easily, just start Spring Boot with `spring.jpa.hibernate.ddl-auto=none` and each entity containing a single ArrayType column, and one entity with the already fixed PostgreSQLEnumJdbcType.

**Comment by cgerteis** [comment]
FYI: I've tried to use `UUID[]` directly instead of `@Type(UUIDArrayType.class)`, but in this particular case it does not seem to work. I also don't see a test for this type in the Hibernate Core tests.

**Comment by vladmihalcea** [comment]
Then, those Types should be added in the Hibernate core. In the long run, that's going to be the best approach since there's support for ARRAYS in Hibernate 6.

**Comment by dinukumar87** [comment]
Hello,
I'm trying to update a object which is having a Enum field, ends with error. below is the detail. inset is fine. i'm using Hibernate 6.4.1 -final and spring 3x. any workarounds will be appreciated. Thanks

```
@Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(table = "", name = "", columnDefinition = "my_enum")
    @JdbcType(PostgreSQLEnumJdbcType.class)
    private MyEnum myEnum;
```
Error log
`20:09:38.805 |ERROR [https-jsse-nio-127.0.0.1-4443-exec-5] o.h.engine.jdbc.spi.SqlExceptionHelper - ERROR: type "myEnum" does not exist
  Position: 447`

Generated SQL
  Position: 447] [merge into table as t using (select cast(? as integer) id, cast(? as boolean) flag, cast(? as text) created_by, cast(? as timestamp(6)) created_ts, cast(? as text) description, cast(? as MyEnum) my_enum) as s on (t.id=s.id) when not matched then insert (id, flag, created_by, created_ts, description,  my_enum) values (s.id, s.flag, s.created_by, s.created_ts, s.description, s.my_enum ....

**Comment by vladmihalcea** [comment]
@dinukumar87 Thats a Hibernate issue, and it shall be asked on either StackOverflow or the Hibernate forum. This issue is about adding a new module for Hibernate 6.4, and it's hard to follow the discussion if we need to skip over various non-related topics that happened to be asked in this issue.

**Comment by vladmihalcea** [comment]
@adrianAzoitei I added the replicated test case on [HHH-17588](https://hibernate.atlassian.net/browse/HHH-17588).

**Comment by reneleonhardt** [comment]
> @reneleonhardt Try to use the Hibernate 6 native array type instead. Check out the [Hibernate Core](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/test/java/org/hibernate/orm/test/type/DateArrayTest.java#L194) test cases and see if they work. If they do, then I will remove the existing ArrayTypes from the 6.4 module.

Pure Int/String Arrays work with Postgres and Hibernate 6.4, but those are no Collections ;)
I hope they will fix their strange breaking changes, no single reactions for 3 weeks to HHH-17588, I don't understand why they don't want to support the most basic Java classes, no one is using Arrays in Java :/

**Comment by reneleonhardt** [comment]
> @reneleonhardt Try to use the Hibernate 6 native array type instead. Check out the [Hibernate Core](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/test/java/org/hibernate/orm/test/type/DateArrayTest.java#L194) test cases and see if they work. If they do, then I will remove the existing ArrayTypes from the 6.4 module.

Thank you, Int/String Arrays work with Postgres and Hibernate 6.4 when I'm using Lists without annotations in Entities and Repos.

The findBy...In methods seem to work too, but only if the database array contains a single value, then I can pass a list of possible search values. But as soon as the database array contains multiple values I can't find a subset by passing one or all search values (in a native query I would use `@>` instead of `IN`). Guess JPA has a long way to go for proper Array/Collection support 😅

**Comment by vladmihalcea** [comment]
@reneleonhardt You will need to create replicating test cases and create Hibernate issues for those findings.

It's good that Hibernate has built-in support for PostgreSQL arrays. If thw HHH-17588issue is not fixed, I will remove the ArrayTypes and release the 64 module.

**Comment by nnrudakov** [comment]
It shows `'Basic' attribute type should not be a container` when using native types in Kotlin with Spring Boot 3.2.1 and Hibernate 6.4.1

**Comment by vladmihalcea** [comment]
@nnrudakov That's an IDE issue and you should reported there. If the project is built fine with Maven or Gradle from the command line, then the warning is just sn IDE problem.

**Comment by nnrudakov** [comment]
> @nnrudakov That's an IDE issue and you should reported there. If the project is built fine with Maven or Gradle from the command line, then the warning is just sn IDE problem.

Yes, thank you. Build complete successfully.

**Comment by marschall** [comment]
`BatchSequenceGenerator` will need a small update, see 

- https://github.com/marschall/hibernate-batch-sequence-generator/issues/15

- https://github.com/marschall/hibernate-batch-sequence-generator/commit/9105cad6a6f818f4c04dbff149023b628e5bcd56#diff-d6fe3ef58e85bda25b8810dc4ae1cf50818d07d4c33ecd4d1403b92b30a469e4

**Comment by vladmihalcea** [comment]
Thanks. I'll merge it once we have the 6.4 module in place.

**Comment by adrianAzoitei** [comment]
Thanks for picking it up and fixing it @vladmihalcea. Did you intentionally keep the module named `-63`?

**Comment by vladmihalcea** [comment]
@adrianAzoitei Knowing that the JSON Type works find for a `List` property `Type`, I figured out that if we change the JDBC Type from `ARRAY` to `OTHER`, the `ArrayType` would work just fine for both 6.3 and 6.4, so there was no reason to create a new module as the `hypersistence-utils-hibernate-63` module can work with both Hibernate 6.3 and 6.4.

**Comment by markusheiden** [comment]
Using `OTHER` breaks Spring Data JPA repository queries against columns that use such a user type.

The "native" solution would be to use Hibernate's builtin support for mapping to SQL arrays:
https://docs.jboss.org/hibernate/orm/6.4/userguide/html_single/Hibernate_User_Guide.html#basic-mapping-array
But currently, Spring Data JPA does not support that:
https://github.com/spring-projects/spring-data-jpa/issues/3356

**Comment by vladmihalcea** [comment]
@markusheiden Unless the [HHH-17588](https://hibernate.atlassian.net/browse/HHH-17588) Hibernate issue is fixed, it's impossible to create a custom `UserType` that maps a `Types.ARRAY`.

The only alternative is the native array Type. In the long run, I suppose it's better to have the built-in Hibernate Types support everything that this library used to provide for arrays.

**Comment by alarv** [comment]
For any poor soul that may be lurking in this topic I'll share my experience, maybe it helps someone:

As mentioned in this thread, custom array types cannot be used with hibernate 6.4, I was using spring boot 3.2.5, which came with hibernate 6.4.4 IIRC. Hibernate `@Array` annotation didn't work. Also, the custom types provided by this library didn't work either. In my case I wanted to map a FloatArray (Float[]) that is mapped in postgresql as float8[]

I upgraded to spring boot 3.3.4 which comes with hibernate 6.5.3.Final and all the problems magically went away. Now I'm mapping my fields as such:
```kotlin
val recall: FloatArray?,
    val f1Score: FloatArray?,
```

and they're being mapped correctly to the database without adding any annotation.
